File: src/hayulo/lexer.py
from __future__ import annotations

from dataclasses import dataclass

from .diagnostics import Diagnostic, HayuloSyntaxError


KEYWORDS = {
    "module",
    "intent",
    "fn",
    "pub",
    "return",
    "if",
    "else",
    "for",
    "in",
    "true",
    "false",
    "and",
    "or",
    "not",
    "test",
    "expect",
}

SINGLE_CHAR_TOKENS = {
    "(": "LPAREN",
    ")": "RPAREN",
    "{": "LBRACE",
    "}": "RBRACE",
    "[": "LBRACKET",
    "]": "RBRACKET",
    ",": "COMMA",
    ":": "COLON",
    ".": "DOT",
    "+": "PLUS",
    "-": "MINUS",
    "*": "STAR",
    "/": "SLASH",
    "%": "PERCENT",
    "=": "EQUAL",
    "<": "LT",
    ">": "GT",
}


@dataclass(frozen=True)
class Token:
    kind: str
    value: str
    line: int
    column: int


class Lexer:
    def __init__(self, source: str, filename: str | None = None):
        self.source = source
        self.filename = filename
        self.index = 0
        self.line = 1
        self.column = 1
        self.tokens: list[Token] = []

    def lex(self) -> list[Token]:
        while not self._is_at_end():
            ch = self._peek()

            if ch in " \t\r":
                self._advance()
                continue

            if ch == "\n":
                self._advance_line()
                continue

            if ch == "/" and self._peek_next() == "/":
                self._skip_comment()
                continue

            if ch == '"':
                self._string()
                continue

            if ch.isdigit():
                self._number()
                continue

            if ch.isalpha() or ch == "_":
                self._identifier()
                continue

            self._symbol()

        self.tokens.append(Token("EOF", "", self.line, self.column))
        return self.tokens

    def _symbol(self) -> None:
        line, col = self.line, self.column
        ch = self._advance()

        if ch == "-" and self._match(">"):
            self.tokens.append(Token("ARROW", "->", line, col))
            return

        if ch == "=" and self._match("="):
            self.tokens.append(Token("EQEQ", "==", line, col))
            return

        if ch == "!" and self._match("="):
            self.tokens.append(Token("BANGEQ", "!=", line, col))
            return

        if ch == "<" and self._match("="):
            self.tokens.append(Token("LTE", "<=", line, col))
            return

        if ch == ">" and self._match("="):
            self.tokens.append(Token("GTE", ">=", line, col))
            return

        kind = SINGLE_CHAR_TOKENS.get(ch)
        if kind is None:
            raise HayuloSyntaxError(
                Diagnostic(
                    code="unexpected_character",
                    message=f"Unexpected character {ch!r}.",
                    file=self.filename,
                    line=line,
                    column=col,
                    suggestions=["Remove this character or replace it with valid Hayulo syntax."],
                )
            )
        self.tokens.append(Token(kind, ch, line, col))

    def _string(self) -> None:
        start_line, start_col = self.line, self.column
        self._advance()  # opening quote
        chars: list[str] = []

        while not self._is_at_end() and self._peek() != '"':
            ch = self._advance()
            if ch == "\\":
                if self._is_at_end():
                    break
                escaped = self._advance()
                mapping = {"n": "\n", "t": "\t", '"': '"', "\\": "\\"}
                chars.append(mapping.get(escaped, escaped))
            elif ch == "\n":
                self.line += 1
                self.column = 1
                chars.append("\n")
            else:
                chars.append(ch)

        if self._is_at_end():
            raise HayuloSyntaxError(
                Diagnostic(
                    code="unterminated_string",
                    message="Unterminated string literal.",
                    file=self.filename,
                    line=start_line,
                    column=start_col,
                    suggestions=["Add a closing double quote."],
                )
            )

        self._advance()  # closing quote
        self.tokens.append(Token("STRING", "".join(chars), start_line, start_col))

    def _number(self) -> None:
        start_line, start_col = self.line, self.column
        chars = [self._advance()]
        is_float = False

        while not self._is_at_end() and self._peek().isdigit():
            chars.append(self._advance())

        if not self._is_at_end() and self._peek() == "." and self._peek_next().isdigit():
            is_float = True
            chars.append(self._advance())
            while not self._is_at_end() and self._peek().isdigit():
                chars.append(self._advance())

        kind = "FLOAT" if is_float else "INT"
        self.tokens.append(Token(kind, "".join(chars), start_line, start_col))

    def _identifier(self) -> None:
        start_line, start_col = self.line, self.column
        chars = [self._advance()]
        while not self._is_at_end() and (self._peek().isalnum() or self._peek() in "_"):
            chars.append(self._advance())
        value = "".join(chars)
        kind = value.upper() if value in KEYWORDS else "IDENT"
        self.tokens.append(Token(kind, value, start_line, start_col))

    def _skip_comment(self) -> None:
        while not self._is_at_end() and self._peek() != "\n":
            self._advance()

    def _advance(self) -> str:
        ch = self.source[self.index]
        self.index += 1
        self.column += 1
        return ch

    def _advance_line(self) -> None:
        self.index += 1
        self.line += 1
        self.column = 1

    def _match(self, expected: str) -> bool:
        if self._is_at_end() or self.source[self.index] != expected:
            return False
        self.index += 1
        self.column += 1
        return True

    def _peek(self) -> str:
        if self._is_at_end():
            return "\0"
        return self.source[self.index]

    def _peek_next(self) -> str:
        if self.index + 1 >= len(self.source):
            return "\0"
        return self.source[self.index + 1]

    def _is_at_end(self) -> bool:
        return self.index >= len(self.source)

File: src/lexer.rs
use crate::diagnostics::{Diagnostic, HayuloError, HayuloResult};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Token {
    pub kind: String,
    pub value: String,
    pub line: usize,
    pub column: usize,
}

pub fn lex(source: &str, filename: Option<&str>) -> HayuloResult<Vec<Token>> {
    Lexer::new(source, filename).lex()
}

struct Lexer<'a> {
    chars: Vec<char>,
    index: usize,
    line: usize,
    column: usize,
    filename: Option<&'a str>,
    tokens: Vec<Token>,
}

impl<'a> Lexer<'a> {
    fn new(source: &str, filename: Option<&'a str>) -> Self {
        Self {
            chars: source.chars().collect(),
            index: 0,
            line: 1,
            column: 1,
            filename,
            tokens: Vec::new(),
        }
    }

    fn lex(mut self) -> HayuloResult<Vec<Token>> {
        while !self.is_at_end() {
            let ch = self.peek();
            if matches!(ch, ' ' | '\t' | '\r') {
                self.advance();
                continue;
            }
            if ch == '\n' {
                self.advance_line();
                continue;
            }
            if ch == '/' && self.peek_next() == '/' {
                self.skip_comment();
                continue;
            }
            if ch == '"' {
                self.string()?;
                continue;
            }
            if ch == '?' {
                return Err(HayuloError::new(
                    Diagnostic::new(
                        "syntax.postfix_try_removed",
                        "Hayulo 2.0 uses prefix try instead of postfix ?.",
                    )
                    .file(self.filename)
                    .line(Some(self.line))
                    .column(Some(self.column))
                    .suggestion("Use `try expression` instead of `expression?`."),
                ));
            }
            if ch.is_ascii_digit() {
                self.number();
                continue;
            }
            if ch.is_ascii_alphabetic() || ch == '_' {
                self.identifier();
                continue;
            }
            self.symbol()?;
        }
        self.tokens.push(Token {
            kind: "EOF".to_string(),
            value: String::new(),
            line: self.line,
            column: self.column,
        });
        Ok(self.tokens)
    }

    fn symbol(&mut self) -> HayuloResult<()> {
        let line = self.line;
        let column = self.column;
        let ch = self.advance();

        if ch == '-' && self.matches('>') {
            self.push("ARROW", "->", line, column);
            return Ok(());
        }
        if ch == '=' && self.matches('=') {
            self.push("EQEQ", "==", line, column);
            return Ok(());
        }
        if ch == '=' && self.matches('>') {
            self.push("FAT_ARROW", "=>", line, column);
            return Ok(());
        }
        if ch == '!' && self.matches('=') {
            self.push("BANGEQ", "!=", line, column);
            return Ok(());
        }
        if ch == '<' && self.matches('=') {
            self.push("LTE", "<=", line, column);
            return Ok(());
        }
        if ch == '>' && self.matches('=') {
            self.push("GTE", ">=", line, column);
            return Ok(());
        }

        let kind = match ch {
            '(' => "LPAREN",
            ')' => "RPAREN",
            '{' => "LBRACE",
            '}' => "RBRACE",
            '[' => "LBRACKET",
            ']' => "RBRACKET",
            ',' => "COMMA",
            ':' => "COLON",
            '.' => "DOT",
            '+' => "PLUS",
            '-' => "MINUS",
            '*' => "STAR",
            '/' => "SLASH",
            '%' => "PERCENT",
            '=' => "EQUAL",
            '<' => "LT",
            '>' => "GT",
            _ => {
                return Err(HayuloError::new(
                    Diagnostic::new(
                        "unexpected_character",
                        format!("Unexpected character {ch:?}."),
                    )
                    .file(self.filename)
                    .line(Some(line))
                    .column(Some(column))
                    .suggestion("Remove this character or replace it with valid Hayulo syntax."),
                ));
            }
        };
        self.push(kind, ch.to_string(), line, column);
        Ok(())
    }

    fn string(&mut self) -> HayuloResult<()> {
        let start_line = self.line;
        let start_col = self.column;
        self.advance();
        let mut value = String::new();
        while !self.is_at_end() && self.peek() != '"' {
            let ch = self.advance();
            if ch == '\\' {
                if self.is_at_end() {
                    break;
                }
                let escaped = self.advance();
                value.push(match escaped {
                    'n' => '\n',
                    't' => '\t',
                    '"' => '"',
                    '\\' => '\\',
                    other => other,
                });
            } else if ch == '\n' {
                self.line += 1;
                self.column = 1;
                value.push('\n');
            } else {
                value.push(ch);
            }
        }
        if self.is_at_end() {
            return Err(HayuloError::new(
                Diagnostic::new("unterminated_string", "Unterminated string literal.")
                    .file(self.filename)
                    .line(Some(start_line))
                    .column(Some(start_col))
                    .suggestion("Add a closing double quote."),
            ));
        }
        self.advance();
        self.push("STRING", value, start_line, start_col);
        Ok(())
    }

    fn number(&mut self) {
        let start_line = self.line;
        let start_col = self.column;
        let mut value = String::new();
        value.push(self.advance());
        while !self.is_at_end() && self.peek().is_ascii_digit() {
            value.push(self.advance());
        }
        let mut kind = "INT";
        if !self.is_at_end() && self.peek() == '.' && self.peek_next().is_ascii_digit() {
            kind = "FLOAT";
            value.push(self.advance());
            while !self.is_at_end() && self.peek().is_ascii_digit() {
                value.push(self.advance());
            }
        }
        self.push(kind, value, start_line, start_col);
    }

    fn identifier(&mut self) {
        let start_line = self.line;
        let start_col = self.column;
        let mut value = String::new();
        value.push(self.advance());
        while !self.is_at_end() && (self.peek().is_ascii_alphanumeric() || self.peek() == '_') {
            value.push(self.advance());
        }
        let kind = match value.as_str() {
            "module" | "intent" | "fn" | "pub" | "let" | "set" | "return" | "try" | "match"
            | "if" | "else" | "for" | "in" | "true" | "false" | "and" | "or" | "not" | "test"
            | "expect" => value.to_ascii_uppercase(),
            _ => "IDENT".to_string(),
        };
        self.push(kind, value, start_line, start_col);
    }

    fn skip_comment(&mut self) {
        while !self.is_at_end() && self.peek() != '\n' {
            self.advance();
        }
    }

    fn push(
        &mut self,
        kind: impl Into<String>,
        value: impl Into<String>,
        line: usize,
        column: usize,
    ) {
        self.tokens.push(Token {
            kind: kind.into(),
            value: value.into(),
            line,
            column,
        });
    }

    fn advance(&mut self) -> char {
        let ch = self.chars[self.index];
        self.index += 1;
        self.column += 1;
        ch
    }

    fn advance_line(&mut self) {
        self.index += 1;
        self.line += 1;
        self.column = 1;
    }

    fn matches(&mut self, expected: char) -> bool {
        if self.is_at_end() || self.chars[self.index] != expected {
            return false;
        }
        self.index += 1;
        self.column += 1;
        true
    }

    fn peek(&self) -> char {
        self.chars.get(self.index).copied().unwrap_or('\0')
    }

    fn peek_next(&self) -> char {
        self.chars.get(self.index + 1).copied().unwrap_or('\0')
    }

    fn is_at_end(&self) -> bool {
        self.index >= self.chars.len()
    }
}
